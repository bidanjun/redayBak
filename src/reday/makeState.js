import React, { Component } from 'react'

// default store,you could made other store same way
export const store = {};

//register state of a component to store
const registerThis = (comp,stateName, storeObejct = store)=>{  
  comp.setState = comp.setState.bind(comp)
  comp.state.setState = comp.setState;
  Object.defineProperty(storeObejct, stateName, {
    get: () => {
      return this.state;
    }
  });
}

//an hoc to create state for wrappedComponent
//most times we use this to make state,and access anywhere by store
// mapFunc could bind some function of component to state : (comp)=>{
// comp.state.func=comp.func.bind(comp);
// comp.state.anotherFunc=comp.anotherFunc.bind(comp)};
//so we could use these function outside component
export default (initialState, stateName, storeObejct = store,mapFunc = null) => WrappedComponent => {
  class State extends Component {
    constructor(props, context) {
      super(props, context);
      this.state = initialState ? initialState : {}
      if (!!mapFunc) mapFunc(this);
      registerThis(this,stateName,storeObejct)
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
        />
      )
    }
  }
  State.displayName = `State(${WrappedComponent.name || WrappedComponent.displayName})`;
  return State
}


// get the state of a component,and register to store
export const registerState= (stateName,storeObjet=store,mapFunc=null)=>(WrappedComponent)=> {
  class Register extends WrappedComponent {
    constructor(props,context) {
      super(props, context);
      if (!!mapFunc) mapFunc(this);
      registerThis(this,stateName,storeObjet)
    }
    render() {
      return super.render()
    }
  }
  Register.displayName = `Register(${WrappedComponent.name || WrappedComponent.displayName})`;
  return Register
}

