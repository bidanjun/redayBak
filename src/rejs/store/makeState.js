
// 自动的为状态增加函数，比如state.counter增加setCounter函数
// 这里同时将属性、状态和改变状态的函数，作为属性传递下去。
// from https://github.com/dai-shi/react-compose-state
import React from 'react';

const isFunction = fn => (typeof fn === 'function');
const capitalize = str => (str.charAt(0).toUpperCase() + str.slice(1));

const composeWithState = initialState => BaseComponent => (
  class extends React.PureComponent {
    constructor(props) {
      super(props);
      if (isFunction(initialState)) {
        this.state = initialState(props);
      } else {
        this.state = initialState;
      }
      this.state = this.state || {};
      this.stateSetters = {};
      Object.keys(this.state).forEach((key) => {
        this.stateSetters[`set${capitalize(key)}`] = (val) => {
          this.setState({ [key]: val });
        };
      });
    }
    render() {
      return (<BaseComponent {...this.props} {...this.state} {...this.stateSetters} />);
    }
  }
);

export default composeWithState;