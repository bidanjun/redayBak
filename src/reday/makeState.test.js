import React from 'react'
import Enzyme, { mount } from 'enzyme'
import serializer from 'enzyme-to-json/serializer'
import Adapter from 'enzyme-adapter-react-16'

import makeState, { store } from './makeState'

Enzyme.configure({ adapter: new Adapter() })
expect.addSnapshotSerializer(serializer)

const increment = state => ({ counter: state.counter + 1 })
test('state should be update when click', () => {
  const wrapper = mount(
    <div>
    {makeState({ counter: 0 }, 'Counter', store)(
      () => (
        <button onClick={() => store.Counter.setState(increment)}>{store.Counter.counter}</button>
      )
    )}
    </div>
  )
  expect(wrapper).toMatchSnapshot(`counte=0`)
  wrapper.find('button').simulate('click')
  expect(wrapper).toMatchSnapshot(`after click,counter=1`)
})