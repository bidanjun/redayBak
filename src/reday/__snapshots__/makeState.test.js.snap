// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`after click,counter=1,result is: 1`] = `<comp />`;

exports[`before click ,counte is 0,result is: 1`] = `<comp />`;
